### ISSUE TITLE
change(next) confuses users: it does not skip dead pets

### ISSUE THREAD

**Issue by bloerwald** [axc450/pbs#36]
Given the script `change(next)`, currently the behavior is as follows

| pet 1 | pet 2 | pet 3 | active | result | 
| -- | -- | -- | -- | -- |
| :green_circle: | :green_circle: | :green_circle: | 1 | :arrow_right: 2 |
| :green_circle: | :green_circle: | :green_circle: | 2 | :arrow_right: 3 |
| :green_circle: | :green_circle: | :green_circle: | 3 | :arrow_right: 1 |
| :green_circle: | :green_circle: | :x: | 1 | :arrow_right: 2 |
| :green_circle: | :green_circle: | :x: | 2 | :cyclone: freeze |
| :green_circle: | :green_circle: | :x: | 3 | :arrow_right: 1 |
| :green_circle: | :x: | :green_circle: | 1 | :cyclone: freeze |
| :green_circle: | :x: | :green_circle: | 2 | :arrow_right: 3 |
| :green_circle: | :x: | :green_circle: | 3 | :arrow_right: 1 |
| :x: | :green_circle: | :green_circle: | 1 | :arrow_right: 2 |
| :x: | :green_circle: | :green_circle: | 2 | :arrow_right: 3 |
| :x: | :green_circle: | :green_circle: | 3 | :cyclone: freeze |
| :x: | :x: | :green_circle: | 1 | :cyclone: freeze |
| :x: | :x: | :green_circle: | 2 | :arrow_right: 3 |
| :x: | :x: | :green_circle: | 3 | :snowflake: freeze |
| :x: | :green_circle: | :x: | 1 | :arrow_right: 2 |
| :x: | :green_circle: | :x: | 2 | :snowflake: freeze |
| :x: | :green_circle: | :x: | 3 | :cyclone: freeze |
| :green_circle: | :x: | :x: | 1 | :snowflake: freeze |
| :green_circle: | :x: | :x: | 2 | :cyclone: freeze |
| :green_circle: | :x: | :x: | 3 | :arrow_right: 1 |

The results marked :cyclone: are confusing to me and at least one user. Our expectation would have been 

| pet 1 | pet 2 | pet 3 | active | result | 
| -- | -- | -- | -- | -- |
| :green_circle: | :green_circle: | :x: | 2 | :arrow_right: 1 |
| :green_circle: | :x: | :green_circle: | 1 | :arrow_right: 3 |
| :x: | :green_circle: | :green_circle: | 3 | :arrow_right: 2 |
| :x: | :x: | :green_circle: | 1 | :arrow_right: 3 |
| :x: | :green_circle: | :x: | 3 | :arrow_right: 2 |
| :green_circle: | :x: | :x: | 2 | :arrow_right: 1 |

i.e. "skip to the next **alive** pet", not "skip to the next pet **if it is alive**".

The current implementation is a dumb `index = index % C_PetBattles.GetNumPets(Enum.BattlePetOwner.Ally) + 1`, so it is working exactly as intended, but I claim that's not what users expect.

There are two options here: Change the behavior of `change(next)`, potentially breaking some user script (but I **really** doubt people actually expect and rely on this behavior). The alternative is to introduce `change(nextusable)` which would check whether the new choice actually *can* be used, only freezing if there is no alternative.

The implementation for either should look somewhat like 

```lua
local function nextAfter(index)
    return index % C_PetBattles.GetNumPets(Enum.BattlePetOwner.Ally) + 1
end
local function canUse(index)
    return C_PetBattles.GetHealth(Enum.BattlePetOwner.Ally, index) ~= 0 and
           C_PetBattles.CanPetSwapIn(index)
end

index = nextAfter(active)
while not canUse(index) and index ~= active do
    index = nextAfter(index)
end
```

The `and C_PetBattles.CanPetSwapIn(index)` condition already exists within this action, but as "if that's the case, don't actually change".  It is up for debate whether the pets `active, forced-to-not-be-active, alive` should switch to `alive` or fail as `forced-to-not-be-active` is alive but unusable (the current behavior). 

I claim that 99% of `change(next)` are either in scripts just mindlessly using a filler pet, or in scripts where the expectation is that one pet dies one after another 1, 2, win style.

I'd be happy with both options, but feel that there is no need to keep the old behavior (and educating people is harder than the effects of breaking it).

**Comment by axc450** [comment]
I think the former implementation is correct here.

I can inform those in the pet community (+ xufus) that this may potentially break some scripts, but I agree it shouldn't be an issue.

**Comment by axc450** [comment]
Released in v1.7
